AppCenter/AppCenterTests/MSAppCenterTests.m

```diff
@@ -975,4 +975,106 @@ - (void)testNoUserIdWhenSetUserIdIsNotCalledInNextVersion {
   XCTAssertNil([[MSUserIdContext sharedInstance] userIdAt:[NSDate dateWithTimeIntervalSince1970:5000]]);
 }
 
+- (void)testServiceNotificationNoRegisteredDelegate {
+  
+  // If
+  // Notification data.
+  NSDictionary<NSString *, NSString *> *data = @{@"key" : @"value"};
+
+  // When
+  [self.sut forwardServiceNotification:data];
+
+  // Just checking this doesn't crash.
+}
+
+- (void)testServiceNotificationAddDelegate {
+
+  // If
+  // Creating delegates.
+  id delegate1 = OCMProtocolMock(@protocol(MSServiceNotificationDelegate));
+  id delegate2 = OCMProtocolMock(@protocol(MSServiceNotificationDelegate));
+
+  // Registering delegates.
+  [self.sut addServiceNotificationDelegate:delegate1];
+  [self.sut addServiceNotificationDelegate:delegate2];
+
+  // Notification data.
+  NSDictionary<NSString *, NSString *> *data = @{@"key" : @"value"};
+
+  // When
+  [self.sut forwardServiceNotification:data];
+
+  // Then
+  OCMVerify([delegate1 appCenter:_sut didReceiveServiceNotification:data]);
+  OCMVerify([delegate2 appCenter:_sut didReceiveServiceNotification:data]);
+}
+
+- (void)testServiceNotificationRemoveAllDelegate {
+
+  // If
+  // Creating delegates.
+  id delegate1 = OCMProtocolMock(@protocol(MSServiceNotificationDelegate));
+  id delegate2 = OCMProtocolMock(@protocol(MSServiceNotificationDelegate));
+
+  // Registering delegates.
+  [self.sut addServiceNotificationDelegate:delegate1];
+  [self.sut addServiceNotificationDelegate:delegate2];
+
+  // Unregistering delegates.
+  [self.sut removeServiceNotificationDelegate:delegate1];
+  [self.sut removeServiceNotificationDelegate:delegate2];
+
+  // Notification data.
+  NSDictionary<NSString *, NSString *> *data = @{@"key" : @"value"};
+
+  // When
+  [self.sut forwardServiceNotification:data];
+
+  // Then
+  OCMReject([delegate1 appCenter:_sut didReceiveServiceNotification:data]);
```

Comment:
The rejects here should go line 1018 after assigning the mocked delegates. Reject actually crashes the test at the time the rejected method is called. This is why it should be set early. Then if the test simply doesn't crash then it succeeds.